sockeye/data_io.py

```diff
@@ -848,25 +847,80 @@ def get_training_data_iters(sources: List[str],
                                     batch_size=batch_size,
                                     bucket_batch_sizes=bucket_batch_sizes,
                                     num_factors=len(sources),
-                                    permute=permute)
-
-    validation_iter = None
-    if validation_sources is not None and validation_target is not None:
-        validation_iter = get_validation_data_iter(data_loader=data_loader,
-                                                   validation_sources=validation_sources,
-                                                   validation_target=validation_target,
-                                                   buckets=buckets,
-                                                   bucket_batch_sizes=bucket_batch_sizes,
-                                                   source_vocabs=source_vocabs,
-                                                   target_vocab=target_vocab,
-                                                   max_seq_len_source=max_seq_len_source,
-                                                   max_seq_len_target=max_seq_len_target,
-                                                   batch_size=batch_size,
-                                                   fill_up=fill_up)
+                                    permute=True)
+
+    validation_iter = get_validation_data_iter(data_loader=data_loader,
+                                               validation_sources=validation_sources,
+                                               validation_target=validation_target,
+                                               buckets=buckets,
+                                               bucket_batch_sizes=bucket_batch_sizes,
+                                               source_vocabs=source_vocabs,
+                                               target_vocab=target_vocab,
+                                               max_seq_len_source=max_seq_len_source,
+                                               max_seq_len_target=max_seq_len_target,
+                                               batch_size=batch_size,
+                                               fill_up=fill_up)
 
     return train_iter, validation_iter, config_data, data_info
 
 
+def get_scoring_data_iters(sources: List[str],
```

Comment:
As suggested in my previous comment: the main contents of this function before creating the `BatchedParallel...Iter` could be part of the constructor of this class itself.

Reply:
This is a good idea, though for posterity, I would like to record here [the Lord of the Rings allusion](http://tolkiengateway.net/wiki/Ring_Verse) I buried in the comments, which will be sadly lost to this refactoring:

```python
    # One bucket to hold them all,
    bucket = (max_seq_len_source, max_seq_len_target)

    # ...One iterator to raise them,
    sources_sentences, target_sentences = create_sequence_readers(sources, target, source_vocabs, target_vocab)

    # ...One loader to mold them all,
    data_loader = RawParallelDatasetLoader(buckets=[bucket],
                                           eos_id=target_vocab[C.EOS_SYMBOL],
                                           pad_id=C.PAD_ID,
                                           skip_blanks=False)

    # ...and with the model appraise them.
    scoring_iter = BatchedRawParallelSampleIter(data_loader=data_loader,
                                                sources_sentences=sources_sentences,
                                                target_sentences=target_sentences,
                                                bucket=bucket,
                                                batch_size=batch_size,
                                                bucket_batch_size=BucketBatchSize(bucket, batch_size, None),
                                                max_lens=(max_seq_len_source, max_seq_len_target),
                                                fill_up=fill_up,
                                                num_factors=len(sources))
```